editor/inserter/menu.js

```diff
@@ -134,24 +138,26 @@ class InserterMenu extends Component {
 	}
 
 	findByIncrement( blockTypes, increment = 1 ) {
-		// Prepend a fake search block to the list to cycle through.
-		const list = [ { name: 'search' }, ...blockTypes ];
-
-		const currentIndex = findIndex( list, ( blockType ) => this.state.currentFocus === blockType.name );
+		const currentIndex = findIndex( blockTypes, ( blockType ) => this.state.currentFocus === blockType.name );
 		const nextIndex = currentIndex + increment;
-		const highestIndex = list.length - 1;
+		const highestIndex = blockTypes.length - 1;
 		const lowestIndex = 0;
 
 		if ( nextIndex > highestIndex ) {
-			return list[ lowestIndex ].name;
+			return 'search';
 		}
 
 		if ( nextIndex < lowestIndex ) {
-			return list[ highestIndex ].name;
+			return 'search';
 		}
 
 		// Return the name of the next block type.
-		return list[ nextIndex ].name;
+		const blockType = blockTypes[ nextIndex ];
+		if ( this.isDisabledBlock( blockType ) ) {
+			return this.findByIncrement( blockTypes, increment > 0 ? increment + 1 : increment - 1 );
```

Comment:
In this case I might have opted for a simple loop rather than recursion, because the `findIndex` above is non-trivial.

```diff
diff --git a/editor/inserter/menu.js b/editor/inserter/menu.js
index 33680f4b..4b95bb9a 100644
--- a/editor/inserter/menu.js
+++ b/editor/inserter/menu.js
@@ -139,10 +139,21 @@ class InserterMenu extends Component {
 
        findByIncrement( blockTypes, increment = 1 ) {
                const currentIndex = findIndex( blockTypes, ( blockType ) => this.state.currentFocus === blockType.name );
-               const nextIndex = currentIndex + increment;
                const highestIndex = blockTypes.length - 1;
                const lowestIndex = 0;
 
+               let nextIndex = currentIndex;
+               let blockType;
+               do {
+                       nextIndex += increment;
+
+                       // Return the name of the next block type.
+                       blockType = blockTypes[ nextIndex ];
+                       if ( blockType && ! this.isDisabledBlock( blockType ) ) {
+                               return blockType.name;
+                       }
+               } while ( blockType );
+
                if ( nextIndex > highestIndex ) {
                        return 'search';
                }
@@ -150,14 +161,6 @@ class InserterMenu extends Component {
                if ( nextIndex < lowestIndex ) {
                        return 'search';
                }
-
-               // Return the name of the next block type.
-               const blockType = blockTypes[ nextIndex ];
-               if ( this.isDisabledBlock( blockType ) ) {
-                       return this.findByIncrement( blockTypes, increment > 0 ? increment + 1 : increment - 1 );
-               }
-
-               return blockType.name;
        }
 
        findNext( blockTypes ) {
```